feat(curriculum): add backend chapter review (#68431)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/review-back-end-development-and-apis/6a439658bfbe59fcf4b221c7.md

- **Client-side limitations**: Client-side JavaScript has restricted file system access, limited support for complex application logic, and potential security risks (e.g., you must not expose database credentials in client-side code).
- **Browser vs. Node global object**: In the browser, the global object is `window`, which exposes DOM manipulation, cookies, and browser events. In Node the global object is `global`, which exposes built-in modules for files, networking, and OS interaction.
- **DOM access**: The browser provides access to the DOM API, but Node does not. Conversely, Node can access almost all system resources including the file system, the browser cannot.
- **Version control**: You can choose which version of Node runs on your server. You have no control over which browser version your users run.
- **Non-blocking, event-driven architecture**: Node uses a single thread and an event loop that efficiently handles many simultaneous requests and I/O operations without blocking, making it ideal for real-time applications.

> - **Advantages of Node on the back-end**: Single language (JavaScript) across front-end and back-end, non-blocking architecture suited to high concurrency, large npm ecosystem, free and open-source.

Reviewer: Added callbacks to the review.

```suggestion
- **Callbacks**: Functions that define what happens once an asynchronous operation completes. Heavy reliance on them is part of why asynchronous Node code can be harder to read and debug.
- **Advantages of Node on the back-end**: Single language (JavaScript) across front-end and back-end, non-blocking architecture suited to high concurrency, large npm ecosystem, free and open-source.
```

- **Listing all installed Node versions**: `nvm ls`
- **Setting a default Node version (macOS/Linux)**: `nvm alias default 22.20.0`
- **nvm-windows**: A separate project for managing Node versions on Windows (not WSL). Download the installer from the official repository and run `nvm install lts` then `nvm use <version>`.
- **Running a JavaScript file with Node**: `node app.js`

## Working with Node Core Modules

### The `fs` Module

- **Purpose**: Provides methods for creating, reading, writing, appending, and deleting files and directories.
- **Three usage styles**: Callback-based (asynchronous), promise-based (`fs/promises`), and synchronous (`Sync` suffix). Synchronous methods block the event loop and are suited only for scripts.
- **`writeFile()`**: Writes content to a file, creating it if it does not exist:

```js
const fs = require("fs/promises");

async function writeToFile() {
  try {
    await fs.writeFile("article.md", "## Node `fs` Module", "utf8");
    console.log("File written to!");
  } catch (err) {
    console.error("Error writing to file:", err);
  }
}
```

- **`appendFile()`**: Appends content to an existing file without overwriting it.

> - **`readFile()`**: Reads file content. Pass `"utf8"` as the encoding to get a string, and omit it to get a `Buffer`.

Reviewer: ```suggestion
- **`readFile()`**: Reads file content. Pass `"utf8"` as the encoding to get a string, or omit it to get a `Buffer`.
```

```

- **Concurrent commands with `&`**: `"dev": "npm run start:server & npm run start:client"`
- **Passing arguments to a script** (use `--` separator): `npm start -- 8000`; arguments accessible via `process.argv[2]`.

### CommonJS Modules

- **CommonJS**: The original Node.js module system. Loads modules synchronously.
- **Importing** with `require()`:

```js
const { multiply } = require('./math');
```

- **Exporting** with `module.exports`:

```js
function multiply(a, b) { return a * b; }
module.exports = { multiply };
```

- CommonJS is the default when `package.json` has `"type": "commonjs"` (or no `"type"` field), or the file has a `.cjs` extension.

### ES Modules (ESM)

- **ES Modules**: The modern, standardized module format. Loaded asynchronously.
- **Named export**: `export function multiply(a, b) { return a * b; }`
- **Default export**: `export default function multiply(a, b) { return a * b; }`
- **Import**: `import { multiply } from './math.mjs';`
- ESM is used when the file has an `.mjs` extension or `package.json` contains `"type": "module"`.

## Understanding How HTTP, DNS, and TCP/IP Work

### Servers and the Client-Server Model

> 

Reviewer: ```suggestion

- **Client-Server Model**: A distributed system architecture where components are either clients (processes that send requests, e.g., a web browser) or servers.
```

- **Server**: A computer or software program that receives requests from clients and returns data or services over a network.
- **Types of servers**: Web servers (HTML/CSS/assets), application servers (business logic), database servers, mail servers, file servers, proxy servers.

### DNS

> - **DNS (Domain Name System)**: Translates human-readable domain names (e.g., `freecodecamp.org`) into IP addresses, often called the "phone book of the internet".

Reviewer: Reworded a bit:

```suggestion
- **DNS (Domain Name System)**, often called the "phone book of the internet": Translates human-readable domain names (e.g., `freecodecamp.org`) into IP addresses.
```

- **3xx Redirection**: `301 Moved Permanently`, `302 Found`, `304 Not Modified`.
- **4xx Client Errors**: `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`.
- **5xx Server Errors**: `500 Internal Server Error`, `502 Bad Gateway`, `503 Service Unavailable`.

## Understanding the HTTP Request-Response Model

- **HTTP methods**: `GET` (retrieve), `POST` (create), `PUT` (full update), `PATCH` (partial update), `DELETE` (remove).
- **Request-response cycle**: Client sends request → Server validates auth, checks method and URL, queries database → Server builds response (status code + optional body) → Client handles response.

### Assets Returned in HTTP Responses

Every response includes a `Content-Type` header telling the browser how to handle the asset:

- **HTML**: `text/html`
- **CSS**: `text/css`
- **JavaScript**: `application/javascript`
- **JSON**: `application/json`
- **PNG/JPEG/WebP**: `image/png`, `image/jpeg`, `image/webp`
- **MP4/MP3**: `video/mp4`, `audio/mpeg`
- **PDF**: `application/pdf`
- **WOFF2**: `font/woff2`

### HTML Form Submissions

- **`action` attribute**: Specifies the URL where form data is sent.

> - **`method` attribute**: Specifies the HTTP method (`GET` or `POST`). HTML does not natively support `PUT`, `PATCH`, or `DELETE`, use the Fetch API or the `method-override` package in Express.

Reviewer: ```suggestion
- **`method` attribute**: Specifies the HTTP method (`GET` or `POST`). HTML does not natively support `PUT`, `PATCH`, or `DELETE`, use the Fetch API or the `method-override` package in Express instead.
```

- **Web service**: A specific type of API that uses web protocols (HTTP/HTTPS) to exchange data over a network. Every web service is an API, but not every API is a web service.
- **Characteristics**: Runs on a server, enables machine-to-machine communication, language-agnostic, platform-agnostic, uses standard web protocols.
- **REST**: Representational State Transfer, the most widely used web service style. Simple, flexible, works naturally with HTTP.
- **SOAP**: Simple Object Access Protocol, uses XML only, common in enterprise systems needing strict standards.
- **XML-RPC**: Uses XML to encode lightweight requests and responses.

### REST Architecture

- **Statelessness**: Each request is handled independently. The server stores no session state between requests. Every request must include all necessary information (e.g., an auth token on every call).
- **Resources**: Everything is a resource identified by a URL, e.g., `/products` for a collection, `/products/1` for a specific item.
- **HTTP methods in REST**: `GET` (retrieve), `POST` (create), `PUT`/`PATCH` (update), `DELETE` (remove).

> - **Response formats**: JSON is most common but XML and plain text also supported. Clients specify format via the `Accept` header:

Reviewer: ```suggestion
- **Response formats**: JSON is most common but XML and plain text are also supported. Clients specify format via the `Accept` header:
```

## Working with Express

- **Express.js**: A minimal and flexible web framework built on top of Node.js. It simplifies building web servers and APIs with far less boilerplate than the native `http` module.
- **A basic Express server**:

```js
const express = require("express");
const app = express();
const port = 3000;

app.get("/", (req, res) => {
  res.send("Hello World!");
});

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`);
});
```

- **`express()`**: Factory function that creates an Express application instance (`app`).
- **`app.listen(port, callback)`**: Starts the server and begins listening for HTTP requests on the specified port.
- **`req`**: The request object, contains URL, headers, params, body, and query data.
- **`res`**: The response object, used to send data back to the client.
- **Middleware**: Code that runs between receiving a request and sending a response.

## Understanding Routing in Express.js

- **Routing**: Defines how an application responds when a client visits a specific URL with a specific HTTP method.

> - **`app.get/post/put/delete(path, handler)`**: Register handlers for the respective HTTP methods.

Reviewer: ```suggestion
- **`app.get/post/put/delete(path, handler)`**: Registers handlers for the respective HTTP methods.
```

});
```

### Debugging and Logging

- **Built-in debug**: Enable by setting the `DEBUG` environment variable: `DEBUG=express:* node index.js`
- **`morgan`**: A popular third-party HTTP request logger middleware. Logs method, URL, status code, and response time:

```js
const morgan = require('morgan');
app.use(morgan('dev'));
```

### Health Checks and Graceful Shutdowns

- **Health check route**: Returns `200 OK` to signal the app is running. Used by load balancers and container orchestrators:

```js
app.get('/health', (req, res) => {
  res.status(200).send('OK');
});
```

- **`SIGTERM`**: Signal sent when a process should terminate. Handle it to allow in-progress requests to finish:

```js
process.on('SIGTERM', () => {
  server.close(() => {
    console.log('Server closed. Cleaning up...');
  });
});
```

- **`SIGINT`**: Signal sent when pressing Ctrl+C during local development. Handle the same way as `SIGTERM`.
- **Marking as unhealthy before shutdown**: Return a `503` from the health check endpoint during shutdown to stop the load balancer from routing new traffic:

```js
let isShuttingDown = false;

app.get('/health', (req, res) => {
  if (isShuttingDown) {

>     res.status(500).send('Shutting down');

Reviewer: ```suggestion
    res.status(503).send('Shutting down');
```

- **Application-level middleware**: Bound to the entire app via `app.use()`. Runs for every incoming request.
- **Router-level middleware**: Bound to a specific `express.Router()` instance via `router.use()`. Applies only to routes handled by that router.
- **Error-handling middleware**: Identified by four parameters `(err, req, res, next)`. Placed at the bottom of the stack.
- **`express.json()`**: Parses incoming requests with JSON payloads. Makes data available on `req.body`.
- **`express.urlencoded({ extended: true })`**: Parses URL-encoded form data.
- **`express.static('public')`**: Serves static files from a directory.
- **`cors`**: Enables Cross-Origin Resource Sharing. Install with `npm install cors`, use with `app.use(cors())`.
- **`morgan`**: Logs HTTP requests. Use `app.use(morgan('tiny'))`.
- Middleware is applied in the order it is registered with `app.use()`.

## Understanding WebSockets

- **WebSockets**: A protocol that creates a persistent, full-duplex (bidirectional) connection between a client and a server, enabling real-time communication.

> - **Problem with HTTP for real-time**: HTTP requires the client to repeatedly poll the server for new data, so it is slow and inefficient for chat apps, live scores boards, and multiplayer games.

Reviewer: ```suggestion
- **Problem with HTTP for real-time**: HTTP requires the client to repeatedly poll the server for new data, so it is slow and inefficient for chat apps, live scoreboards, and multiplayer games.
```

- **WebSocket handshake**: Every WebSocket connection begins with an HTTP request. The client requests an upgrade and the server responds with `101 Switching Protocols`. After that the connection stays open until either side sends a close signal.
- **Use cases**: Chat applications, online multiplayer games, live sports dashboards, stock market feeds, collaborative tools.

### The `ws` Library

> 

Reviewer: ```suggestion

- Node.js has no built-in WebSocket server; the `ws` library is the common way to implement one.

```

- **Advantages over direct communication**: Decouples components, a publisher does not need to know who its subscribers are.
- **Pub/Sub vs. WebSockets**: WebSockets are the communication channel (the road). Pub/Sub is the messaging pattern that determines who receives which messages (the traffic system).

## Introduction to Authentication and Authorization

### Authentication vs. Authorization

- **Authentication**: Verifies identity, answers the "Who are you?" question, must happen before authorization.
- **Authorization**: Determines permissions, answers the "What are you allowed to do?" question.
- **Three authentication factors**:
  - **Something you know**: Password, PIN, security question answer.
  - **Something you have**: Smartphone receiving an SMS code, authenticator app token.
  - **Something you are**: Fingerprint, facial recognition (biometrics).
- **Common roles in authorization**: Standard User (limited self-service), Editor (content management), Admin (full system access).
- In backend code, authorization is usually enforced via middleware that checks user roles or permissions before granting access.

### Cross-Site Request Forgery (CSRF)

> - **CSRF**: An attack where a malicious website tricks an authenticated user's browser into making an unwanted request to a backend server, exploiting the browser automatic cookie attachment.

Reviewer: ```suggestion
- **CSRF**: An attack where a malicious website tricks an authenticated user's browser into making an unwanted request to a backend server, exploiting the browser's automatic cookie attachment.
```

- **Attack flow**: User logs into `trusted-bank.com` → visits `evil-hacker-site.com` → malicious site submits a hidden form to the bank → bank server sees valid session cookie and executes the request.
- **`SameSite=Strict`**: Browser never sends the cookie if the request originates from a third-party website.
- **`SameSite=Lax`**: Browser allows cookies only when a user safely navigates to the target site (clicking a link), but blocks them on cross-site form submissions.
- **CSRF tokens**: Server generates a random unpredictable token embedded in forms. Client must include the token on every state-changing request. A malicious site cannot read the token from the frontend, so its forged requests will fail validation.

### JSON Web Tokens (JWT)

- **JWT**: A digitally signed token issued by the server after successful login. The client stores it and sends it with every subsequent request, enabling stateless authentication (no server-side session storage needed).

> - **JWT structure**: Three Base64-encoded parts separated by dots:

Reviewer: ```suggestion
- **JWT structure**: Three Base64URL-encoded parts separated by dots:
```

```js
app.post('/login', passport.authenticate('local'), (req, res) => {
  res.send(`Welcome back, ${req.user.username}!`);
});
```

- After successful authentication, Passport stores a piece of user data in the session cookie and attaches the authenticated user object to `req.user`.

### Helmet.js

- **Helmet.js**: A security middleware collection that protects Express apps by automatically setting, modifying, or removing HTTP response headers.

```js
const helmet = require('helmet');
app.use(helmet());
```

- **Removes `X-Powered-By`**: Prevents attackers from knowing the server technology.
- **Sets `Content-Security-Policy`**: Restricts where scripts, images, and styles can be loaded from, mitigating XSS attacks.
- **Sets `X-Frame-Options: SAMEORIGIN`**: Prevents clickjacking by blocking your site from being embedded in external iframes.
- **Sets `Strict-Transport-Security`**: Forces browsers to use HTTPS, protecting data from interception.

## Understanding Security and Privacy in Web Applications

### Security vs. Privacy

- **Security**: The technical defense mechanisms that protect systems and data from unauthorized access. Security provides the lock on the door.

> - **Privacy**: The governance of how personal data is collected, used, and shared. Privacy provides the curtain behind the door.

Reviewer: ```suggestion
- **Privacy**: The governance of how personal data is collected, used, and shared. Privacy provides the curtain on the window.
```

- **Right to be Forgotten (data erasure)**: The legal right to request permanent deletion of all personal records from a company's databases after closing an account.

### The `.env` File and `dotenv`

- **`.env` file**: A plain-text file at the project root that stores configuration secrets such as API keys, database passwords, and port numbers as key-value pairs. Always add it to `.gitignore` to keep secrets out of version control.
- **`dotenv` package**: Reads the `.env` file at startup and injects the values into `process.env`:

```js
require('dotenv').config();

const port = process.env.PORT || 5000;
const apiKey = process.env.API_KEY;
```

- **Two key benefits**: Credentials stay out of source code (security), and the same codebase can behave differently across development, staging, and production by swapping the `.env` file on each server (flexibility).

### Regional Privacy Laws and Compliance

- **GDPR (General Data Protection Regulation)**: Protects individuals in the EU. Requires explicit opt-in consent before tracking, with no pre-ticked boxes allowed. Grants users the right to view and permanently delete their data. Carries heavy financial penalties for violations.

> - **CCPA / CPRA (California Consumer Privacy Act)**: Protects California residents. Focuses on opt-out rights rather than opt-in. Requires a visible "Do Not Sell My Personal Information" link on the homepage.

Reviewer: ```suggestion
- **CCPA (California Consumer Privacy Act) / CPRA (California Privacy Rights Act)**: Protects California residents. Focuses on opt-out rights rather than opt-in. Requires a visible "Do Not Sell My Personal Information" link on the homepage.
```

- **COPPA (Children's Online Privacy Protection Act)**: US federal law protecting children under 13. Requires a clear privacy policy and verifiable parental consent before collecting personal data from a child.
- **DPA (Data Protection Act)**: The UK's principal data privacy framework. Works alongside the UK GDPR and grants users the right to access and delete personal data.
- Other countries have their own laws: Nigeria has the NDPA (Nigeria Data Protection Act) and India has the DPDP Act (Digital Personal Data Protection Act).
- **Compliance steps for global teams**:
  - Build location-aware consent banners that display the appropriate cookie notice based on the user's geographic location.
  - Implement an automated DSAR (Data Subject Access Request) system to find, export, or permanently erase a user's data upon their legal request.
  - Practice data minimization and use age-verification gates to prevent unauthorized tracking of minors.

>   - Designate a Data Protection Officer (DPO) to oversee the privacy architecture and act as the main contact for regulatory authorities.

Reviewer: Assignment is missing at the end:

```md
# --assignment--

Review the Back End Development and APIs topics and concepts.
```

---
id: 6a439658bfbe59fcf4b221c7
title: Back End Development and APIs Review
challengeType: 31
dashedName: review-back-end-development-and-apis
---

# --description--

## Working with Node.js and Event-Driven Architecture

- **Node.js**: An open-source, cross-platform JavaScript runtime environment that allows you to run JavaScript code outside of a browser. It is widely used to build web servers, APIs, and back-end systems.
- **Client-side limitations**: Client-side JavaScript has restricted file system access, limited support for complex application logic, and potential security risks (e.g., you must not expose database credentials in client-side code).

> - **Browser vs. Node global object**: In the browser, the global object is `window`, which exposes DOM manipulation, cookies, and browser events. In Node the global object is `global`, which exposes built-in modules for files, networking, and OS interaction.

Reviewer 2: `In Node` should probably be `In Node.js` to match the earlier heading style

```suggestion
- **Browser vs. Node.js global object**: In the browser, the global object is `window`, which exposes DOM manipulation, cookies, and browser events. In Node the global object is `global`, which exposes built-in modules for files, networking, and OS interaction.
```

> - **DOM access**: The browser provides access to the DOM API, but Node does not. Conversely, Node can access almost all system resources including the file system, the browser cannot.

Reviewer 2: ```suggestion
- **DOM access**: The browser provides access to the DOM API, but Node does not. Conversely, Node can access almost all system resources, including the file system, while the browser cannot.
```
